### PR TITLE
[ENH] centralise and tidy up the converters

### PR DESCRIPTION
**Pull request by TonyBagnall**
This PR has been split into two and is part of a sequence of four. 

1. https://github.com/aeon-toolkit/aeon/pull/606 implements utility functions to test data characteristics and perform conversions, all still built on datatypes. Basically duplicates some functionality currently done by multiple calls to checkisscitype
2. This PR pulls all the conversions for collections into _convert_collections, and it also provides the missing conversions (listed in NOT_IMPLEMENTED_CONVERSIONS)

Note  these are not yet called, it is a set up before revising classifier base then adding a new base class. this PR pulls the converters used by the collections estimators into one place, simplifies things in relation to datatypes and provides comprehensive conversion. Note that conversions of some of the more unusual types are not done as efficiently as they could be, since I have just combined two other conversions. Given we allow seven different data structures and some of them are never used, it seemed reasonable to me, I can create an issue to reimplement after this is in.

Next steps are:

3. Tidying up classification base class to use these 
https://github.com/aeon-toolkit/aeon/pull/551

4. generalise to a common base class inbetween BaseEstimator and BaseClassifier/Regressor/Clusterer/CollectionsTransformer to share the same conversion normalisation. 

#### Reference Issues/PRs

part of #42 see #557

## Did you add any tests for the change?

yes I think I covered it all with tests

#### Any other comments?

Wasnt sure where to put these

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.
- [ ] For new estimators, I've added the estimator to the online documentation.

### CODE REVIEW

File: aeon/utils/validation/_convert_collection.py [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+    X_mi : list
+        list of pd.DataFrame (n_instances, n_timepoints)
+    """
+    if not isinstance(X, np.ndarray) or X.ndim != 2:
+        raise TypeError(numpyflat_error)
+    X_3d = X.reshape(X.shape[0], 1, X.shape[1])
+    X_list = []
+    for x in X_3d:
+        X_list.append(pd.DataFrame(x))
+    return X_list
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/_convert_collection.py:310]: Docstring seems incorrect, not a list of numpy.

It does not look like it is being transposed either? the dataframes should be (n_timepoints,1) right?

File: aeon/utils/validation/_convert_collection.py [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+    X_flat = _from_numpy3d_to_numpyflat(X)
+    return pd.DataFrame(X_flat)
+
+
+def _from_numpy3d_to_numpyflat(X):
+    """Convert 3D np.darray to a 2D np.ndarray."""
+    if not isinstance(X, np.ndarray) and X.ndim != 3:
+        raise TypeError(numpy3D_error)
+    X_flat = X.reshape(X.shape[0], X.shape[1] * X.shape[2])
+    return X_flat
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/_convert_collection.py:117]: This concatenates all of the multivariate series into a single series.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Is this duplicating the other conversion functionality? I do not see other stuff being removed and am wondering if thats just a todo.

File: aeon/utils/validation/_convert_collection.py [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    return _from_numpy3d_to_pd_multiindex(X_3d)
+
+
+def _from_pd_wide_to_numpy3d(X):
+    X = X.to_numpy()
+    return _from_numpyflat_to_numpy3d(X)
+
+
+def _from_pd_wide_to_np_list(X):
+    """Docstring placeholder."""
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/_convert_collection.py:355]: not needed

File: aeon/utils/validation/_convert_collection.py (hunk 1 above)
Comment (MatthewMiddlehurst) [reply aeon/utils/validation/_convert_collection.py:310]: same with other pd_list stuff

**Comment by TonyBagnall** [comment]
> Is this duplicating the other conversion functionality? I do not see other stuff being removed and am wondering if thats just a todo.

it is duplicating in preparation for use in base classifier. I am not going to remove the existing ones to safely eventually deprecate datatypes. My plan is to isolate it first.

File: aeon/utils/validation/_convert_collection.py (hunk 1 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:310]: I cannot find it documented anywhere what shape df_list should be. The only reference to it anywhere is 
    ("df-list", "Panel", "list of pd.DataFrame"). The example does not make it clear

cols = [f"var_{i}" for i in range(1)]
Xlist = [
    pd.DataFrame([[4], [5], [6]], columns=cols),
    pd.DataFrame([[4], [55], [6]], columns=cols),
    pd.DataFrame([[42], [5], [6]], columns=cols),
]

example_dict[("df-list", "Panel", 1)] = Xlist

the only place it is ever used is multirocket_multivariate_variable. where it is transposed, but this class concatenates it all weirdly as far as I can see. So yes, I guess its sole use is (n_timepoints, n_channels). I think it might be easier just to purge it until we need it, np_list much easier to do.

File: aeon/utils/validation/_convert_collection.py (hunk 2 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:117]: he current definition of numpyflat, n_cases, n_channels*n_timepoints, I thought it did that?

File: aeon/utils/validation/_convert_collection.py (hunk 1 above)
Comment (MatthewMiddlehurst) [reply aeon/utils/validation/_convert_collection.py:310]: its in Section 1.2.3 of the big datatypes notebook

I think keep it for now as (n_timepoints, n_channels), probably deserves its own PR if its being changed or removed

Comment (MatthewMiddlehurst) [reply aeon/utils/validation/_convert_collection.py:310]: There was an issue with these converters fixed in #509

File: aeon/utils/validation/_convert_collection.py (hunk 2 above)
Comment (MatthewMiddlehurst) [reply aeon/utils/validation/_convert_collection.py:117]: Where is that definition? I thought I deleted the original converter (#509) because it is not possible to tell where each multivariate series begins and ends.

Creating one big series is not changing the data type, its just transforming the data. I wouldn't mind a parameter to do so, but having the default functionality be that is not good IMO.

File: aeon/utils/validation/_convert_collection.py (hunk 1 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:310]: > its in Section 1.2.3 of the big datatypes notebook
> 
> I think keep it for now as (n_timepoints, n_channels), probably deserves its own PR if its being changed or removed

dont mention the aa datatypes notebook :)

**Comment by TonyBagnall** [comment]
none of this is being used yet, should probably have kept it draft, but good spots all round

**Comment by MatthewMiddlehurst** [comment]
It can still be converted to one if its not ready, unsure at what stage it is at 🙂

File: aeon/utils/validation/_convert_collection.py (hunk 2 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:117]: in the registry. if its univariate only, then would throw an error from the converter if n_channels >1 I guess. None of this is a priority, I'll return this to draft.  

    (
        "numpyflat",
        "Panel",
        "2D np.ndarray of format (n_cases, n_channels*n_timepoints)",
    ),

Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:117]: I am not supporting this usage of numpyflat, although it could be deployed for transformers to allow n_cases, n_timepoints

**Comment by TonyBagnall** [comment]
now the fit overriding classifiers are gone I will flip this around, do the base classifier first then tidy up the conversions I think

File: aeon/utils/validation/_convert_collection.py (hunk 1 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py]: reverted to n_timepoints, n_cases

File: aeon/utils/validation/_convert_collection.py (hunk 2 above)
Comment (MatthewMiddlehurst) [reply aeon/utils/validation/_convert_collection.py:152]: are we leaving this for a different PR?

File: aeon/utils/_testing/scenarios.py [hunk 4]
```diff
@@ -269,7 +269,6 @@ def _check_dict_of_dict(obj, name="obj"):
         and np.all([isinstance(x, str) for x in obj.keys()])
     ):
         raise TypeError(
-            f"Argument {name} with must be a dict of dict, with str keys, "
-            f"found {type(obj)}"
```
Comment (MatthewMiddlehurst) [inline aeon/utils/_testing/scenarios.py:273]: is this change intended? seems very minor but not quite sure of the point

File: aeon/utils/validation/_convert_collection.py [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+        x_nested[_label] = pd.Series(dim_list)
+    x_nested = pd.DataFrame(x_nested).set_axis(instance_idxs)
+
+    col_msg = "Multi-index and nested DataFrames should have same columns names"
+    assert (x_nested.columns == X.columns).all(), col_msg
+
+    return x_nested
+
+
+def _equal_length(X, input_type):
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/_convert_collection.py:564]: any reasoning for this to go from public to private? seems like it could be useful elsewhere

File: aeon/utils/validation/tests/test_collection.py [hunk 6]
```diff
@@ -8,12 +8,14 @@
...
 # from aeon.datasets._data_generators import make_example_multi_index_dataframe
 from aeon.utils._testing.tests.test_collection import make_nested_dataframe_data
-from aeon.utils.validation.collection import (  # _nested_uni_is_equal,; ,
-    COLLECTIONS_DATA_TYPES,
+from aeon.utils.validation._convert_collection import (
+    _equal_length,
     _is_nested_univ_dataframe,
     _is_pd_wide,
+)
+from aeon.utils.validation.collection import (  # _nested_uni_is_equal,; ,
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/tests/test_collection.py:16]: odd comment here, was there previously but may be worth just removing

File: aeon/utils/_testing/scenarios.py (hunk 4 above)
Comment (TonyBagnall) [reply aeon/utils/_testing/scenarios.py:273]: its fixing a problem that was already fixed in a concurrent PR I think

File: aeon/utils/validation/_convert_collection.py (hunk 5 above)
Comment (TonyBagnall) [reply aeon/utils/validation/_convert_collection.py:564]: yes I thought it confusing to have is_equal_length and equal_length both public. Also, _equal_length assumes the user knows valid input type, and I would like to encapsulate that information as much as possible

File: aeon/utils/validation/tests/test_collection.py (hunk 6 above)
Comment (TonyBagnall) [reply aeon/utils/validation/tests/test_collection.py:16]: some legacy comment, Ive removed it :)

**Review by MatthewMiddlehurst — APPROVED** [review]
